src/main/java/epicsquid/roots/entity/ritual/EntityRitualFlowerGrowth.java

```diff
@@ -36,23 +43,53 @@ public void onUpdate() {
   }
 
   private boolean generateFlower(BlockPos pos) {
-    IBlockState flower = getRandomFlower();
-    if (world.isAirBlock(pos) && flower.getBlock().canPlaceBlockAt(world, pos)) {
-      if (!world.isRemote) {
-        world.setBlockState(pos, flower);
-      }
-      return true;
+    if (!world.isAirBlock(pos)) {
+      return false;
     }
-    return false;
-  }
 
-  private IBlockState getRandomFlower() {
+    List<IIngredient> allowedSoils = Collections.emptyList();
+    // If the user removed all recipes, we'll fall back to making dandelions.
+    IBlockState flower = Blocks.YELLOW_FLOWER.getStateFromMeta(BlockFlower.EnumFlowerType.DANDELION.getMeta());
+
+    // Ideally, we'd iterate through all recipes in a random order and try to use each.
+    // That would allow us to place one flower per iteration more reliably instead of trying a random recipe
+    // at every point in a giant column.
     FlowerRecipe recipe = ModRecipes.getRandomFlowerRecipe();
-    if (recipe == null) return Blocks.YELLOW_FLOWER.getStateFromMeta(BlockFlower.EnumFlowerType.DANDELION.getMeta());
+    if (recipe != null) {
```

Comment:
Some of this logic could be abstracted out to `ModRecipes` (which is a nightmare, I know), as you suggest.

Passing the down state as an option and then constructing the ItemStack and comparing it against the allowed ingredients of the randomly picked recipe, picking a new one if it fails, or alternately dropping out and returning null.

This null check on the `getRandomFlowerRecipe` call should prevent getting stuck in an endless loop.

Reply:
Done. The plus side: we'll successfully plant a flower much more often. The downside: the caller is responsible for ensuring the soil block is not air.

Reply:
Correction: MysticalLib handles air correctly.